pkg/ipamd/datastore/data_store_test.go

```diff
@@ -141,6 +192,49 @@ func TestAddENIIPv4Address(t *testing.T) {
 
 }
 
+func TestAddENIIPv4AddressWithPDEnabled(t *testing.T) {
+	ds := NewDataStore(Testlog, NullCheckpoint{}, true)
+
+	err := ds.AddENI("eni-1", 1, true, false, false)
+	assert.NoError(t, err)
+
+	err = ds.AddENI("eni-2", 2, false, false, false)
+	assert.NoError(t, err)
+
+	ipv4Addr := net.IPNet{IP: net.ParseIP("10.0.0.0"), Mask: net.IPv4Mask(255, 255, 255, 240)}
+	err = ds.AddIPv4CidrToStore("eni-1", ipv4Addr, true)
+	assert.NoError(t, err)
+	assert.Equal(t, ds.total, 16)
+	assert.Equal(t, len(ds.eniPool["eni-1"].AvailableIPv4Cidrs), 1)
+
+	ipv4Addr = net.IPNet{IP: net.ParseIP("10.0.0.0"), Mask: net.IPv4Mask(255, 255, 255, 240)}
+	err = ds.AddIPv4CidrToStore("eni-1", ipv4Addr, true)
+	assert.Error(t, err)
+	assert.Equal(t, ds.total, 16)
+	assert.Equal(t, len(ds.eniPool["eni-1"].AvailableIPv4Cidrs), 1)
+
+	ipv4Addr = net.IPNet{IP: net.ParseIP("20.0.0.0"), Mask: net.IPv4Mask(255, 255, 255, 240)}
+	err = ds.AddIPv4CidrToStore("eni-1", ipv4Addr, true)
+	assert.NoError(t, err)
+	assert.Equal(t, ds.total, 32)
+	assert.Equal(t, len(ds.eniPool["eni-1"].AvailableIPv4Cidrs), 2)
+
+	ipv4Addr = net.IPNet{IP: net.ParseIP("30.0.0.0"), Mask: net.IPv4Mask(255, 255, 255, 240)}
+	err = ds.AddIPv4CidrToStore("eni-2", ipv4Addr, true)
+	assert.NoError(t, err)
+	assert.Equal(t, ds.total, 48)
+	assert.Equal(t, len(ds.eniPool["eni-1"].AvailableIPv4Cidrs), 2)
+	assert.Equal(t, len(ds.eniPool["eni-2"].AvailableIPv4Cidrs), 1)
+
+	ipv4Addr = net.IPNet{IP: net.ParseIP("40.0.0.0"), Mask: net.IPv4Mask(255, 255, 255, 240)}
+	err = ds.AddIPv4CidrToStore("dummy-eni", ipv4Addr, true)
+	assert.Error(t, err)
+	assert.Equal(t, ds.total, 48)
+	assert.Equal(t, len(ds.eniPool["eni-1"].AvailableIPv4Cidrs), 2)
+	assert.Equal(t, len(ds.eniPool["eni-2"].AvailableIPv4Cidrs), 1)
+
```

Comment:
nit: can you add a case of adding the same ENI,addr (not the update case) and verify that the operation is idempotent?

Reply:
Line 210, does add the same (ENI,Addr) and expected to error out.